File: Stories.ServiceLayer/Impl/GroupService.cs
using Stories.ServiceLayer.Abstract;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Stories.DataModels.Entities;
using Stories.DataAccessLayer;

namespace Stories.ServiceLayer.Impl
{
    public class GroupService : ServiceBase, IGroupService
    {
        public GroupService(IUnitOfWork unitOfWork) : base(unitOfWork)
        {
        }

        public async Task<Group> CreateGroup(Group group)
        {
            group.Id = await _UnitOfWork.GroupRepository.AddOrAttach(group, true);
            return group;
        }

        public async Task<IList<Group>> GetAllGroups(int userId, int first = 0, int pageSize = 0)
        {
            var groups = await _UnitOfWork.GroupRepository.GetAllGroups(userId, first: first, pageSize: pageSize);
            return groups.ToList();
        }

        public async Task<IList<int>> GetGroupsUsers(IEnumerable<int> groupIds)
        {
            return await _UnitOfWork.SqlQueryAsync<int>($"select distinct UserId from GroupUsers where GroupId in ({string.Join(",", groupIds.Select(x => x.ToString()))})");
        }

        public async Task<int> JoinOrLeaveGroup(int userId, int groupId, bool join)
        {
            await _UnitOfWork.GroupRepository.JoinOrLeaveGroup(userId, groupId, join);
            await _UnitOfWork.SaveAsync();
            return join ? 1 : -1;
        }

        public async Task<IList<Group>> SearchGroup(string searchText)
        {
            return await _UnitOfWork.GroupRepository.GetAll(x => x.Name.Contains(searchText));
        }
    }
}

File: Stories.WebApp/Scripts/app/services/groupService.js
(function () {
    'use strict';
    angular
        .module('myapp')
        .factory('groupService', ['$http', 'appSettings', function ($http, appSettings) {
            var service = {};
            var _loadGroups = function (first, pagesize) {
                return $http.get(appSettings.BaseUrl + 'Group/GetAllGroups?first=' + first + '&pagesize=' + pagesize).then(function (result) {
                    return result;
                });
            };
            var _joinOrLeaveGroup = function (groupId, join) {
                return $http.post(appSettings.BaseUrl + 'Group/joinOrLeaveGroup?groupId=' + groupId + '&join=' + join).then(function (result) {
                    return result;
                });
            };
            var _saveGroup = function (group) {
                return $http.post(appSettings.BaseUrl + 'Group/SaveGroup', group).then(function (result) {
                    return result;
                });
            };
            var _searchGroups = function (searchText) {
                return $http.get(appSettings.BaseUrl + 'Group/SearchGroups?searchText=' +  searchText).then(function (result) {
                    return result;
                });
            };

            service.loadGroups = _loadGroups;
            service.joinOrLeaveGroup = _joinOrLeaveGroup;
            service.saveGroup = _saveGroup;
            service.searchGroups = _searchGroups;
            return service;
        }]);
})();